src/components/__tests__/UserForm.test.tsx

```diff
@@ -0,0 +1,64 @@
+import { render, screen, renderHook, act } from '@testing-library/react';
+import UserForm from '../UserForm';
+import { FormInput } from '../../types';
+import userEvent from '@testing-library/user-event';
+import useStateWithLocalStorage from '../UseStateWithLocalStorage';
+import { debug } from 'console';
+
+const typeIntoForm = ({ name }: FormInput) => {
+  const inputElement = screen.getByRole<HTMLInputElement>('textbox', {
+    name: /name/i,
+  });
+
+  if (name) {
+    userEvent.type(inputElement, name);
+  }
+
+  return {
+    inputElement,
+  };
+};
+
+describe('Input form', () => {
+  beforeEach(() => {
+    render(<UserForm />);
+  });
+
+  test('input should initially be empty', () => {
+    const userInput = screen.getByRole('textbox');
+    expect(userInput.value).toBe('');
+  });
+
+  test('should be able to type a name', () => {
```

Comment:
Test is not testing anything only what you have created on line 8.

I would keep the render inside each test and not in a before each. 

To test I would do something along the lines of this
```javascript
test('should be able to type a name', () => {
  const mockSetState = jest.fn();
  jest.mock("./UseStateWithLocalStorage", () => () => {
    const MockName = "UseStateWithLocalStorage";
    return [jest.fun(), mockSetState];
  });

  const { getByRole } = render(<UserForm />);
  fireEvent.change(getByRole('textbox'), {target: {value: 'Harry'}})
  // Check if the correct data has been sent to the function from the input
  expect(mockSetState).toBeCalledWith()
})
```

Reply:
Might have to play around with this one, It's just off the top of my head

Reply:
for this test I'm a little confused as to why I'd need to use UseStateWithLocalStorage when I'm only testing that the user could type into the input and not testing if the state has changed. I was thinking something possibly like:

```
const setup = () => {
  const utils = render(<UserForm />);
  const input = utils.getByRole('textbox');
  return {
    input,
    ...utils,
  };
};
```
```
test('should be able to type a name', () => {
    const { input } = setup();
    fireEvent.change(input, { target: { value: 'matt' } });
    expect(input.value).toBe('matt');
  });
```

Reply:
The functionality that you've created is not the input it is what happens to the inputs value. if you removed everything in the component except the input the test would (most likely) still pass.

Your second suggestion is cleaner however it is still suffering from the same issue. The reason we mock out `UseStateWithLocalStorage` and check what is being passed to it is because that is the functionality of the app that we have created